From 87790d703798829aa21b0ad89c988e17a7767455 Mon Sep 17 00:00:00 2001
From: David Daney <david.daney@cavium.com>
Date: Fri, 7 Oct 2011 12:01:19 -0700
Subject: [PATCH 161/337] netdev: octeon-ethernet: Report some ethtool
 information for ports without PHYs.

Based On SDK 3.0.0-482

For ports with no connected PHY driver, we can still report some
minimal information about the link.  Do that instead of returning
-EINVAL.

Signed-off-by: David Daney <david.daney@cavium.com>
Signed-off-by: Corey Minyard <cminyard@mvista.com>
Signed-off-by: Jack Tan <jack.tan@windriver.com>
---
 drivers/net/ethernet/octeon/ethernet-mdio.c |   11 ++++++++---
 1 files changed, 8 insertions(+), 3 deletions(-)

diff --git a/drivers/net/ethernet/octeon/ethernet-mdio.c b/drivers/net/ethernet/octeon/ethernet-mdio.c
index 4d00fc8..158ea09 100644
--- a/drivers/net/ethernet/octeon/ethernet-mdio.c
+++ b/drivers/net/ethernet/octeon/ethernet-mdio.c
@@ -55,12 +55,17 @@ static void cvm_oct_get_drvinfo(struct net_device *dev,
 
 static int cvm_oct_get_settings(struct net_device *dev, struct ethtool_cmd *cmd)
 {
+	cvmx_helper_link_info_t link_info;
 	struct octeon_ethernet *priv = netdev_priv(dev);
 
 	if (priv->phydev)
 		return phy_ethtool_gset(priv->phydev, cmd);
 
-	return -EINVAL;
+	link_info = cvmx_helper_link_get(priv->ipd_port);
+	cmd->speed = link_info.s.link_up ? link_info.s.speed : 0;
+	cmd->duplex = link_info.s.full_duplex ? DUPLEX_FULL : DUPLEX_HALF;
+
+	return 0;
 }
 
 static int cvm_oct_set_settings(struct net_device *dev, struct ethtool_cmd *cmd)
@@ -73,7 +78,7 @@ static int cvm_oct_set_settings(struct net_device *dev, struct ethtool_cmd *cmd)
 	if (priv->phydev)
 		return phy_ethtool_sset(priv->phydev, cmd);
 
-	return -EINVAL;
+	return -EOPNOTSUPP;
 }
 
 static int cvm_oct_nway_reset(struct net_device *dev)
@@ -86,7 +91,7 @@ static int cvm_oct_nway_reset(struct net_device *dev)
 	if (priv->phydev)
 		return phy_start_aneg(priv->phydev);
 
-	return -EINVAL;
+	return -EOPNOTSUPP;
 }
 
 const struct ethtool_ops cvm_oct_ethtool_ops = {
-- 
1.7.5.4

